gtsfm/loader/loader_base.py

```diff
@@ -81,6 +95,73 @@ def is_valid_pair(self, idx1: int, idx2: int) -> bool:
             validation result.
         """
 
+    def get_image(self, index: int) -> Image:
+        """Get the image at the given index, satisfying a maximum image resolution constraint.
+
+        Determine how the camera intrinsics and images should be jointly rescaled based on desired img. resolution.
+        Each loader implementation should set a `_max_resolution` attribute.
+
+        Args:
+            index: the index to fetch.
+
+        Raises:
+            IndexError: if an out-of-bounds image index is requested.
+
+        Returns:
+            Image: the image at the query index. It will be resized to satisfy the maximum
+                allowed loader image resolution if the full-resolution images for a dataset
+                are too large.
+        """
+        # no downsampling may be required, in which case target_h and target_w will be identical
+        # to the full res height & width.
+        img_full_res = self.get_image_full_res(index)
+        (
+            _,
+            _,
+            target_h,
+            target_w,
+        ) = img_utils.get_downsampling_factor_per_axis(img_full_res.height, img_full_res.width, self._max_resolution)
+        logger.info(
+            "Image %d resized from (H,W)=(%d,%d) -> (%d,%d)",
+            index,
+            img_full_res.height,
+            img_full_res.width,
+            target_h,
+            target_w,
+        )
+        resized_img = img_utils.resize_image(img_full_res, new_height=target_h, new_width=target_w)
+        return resized_img
+
+    def get_camera_intrinsics(self, index: int) -> Cal3Bundler:
+        """Get the camera intrinsics at the given index, for a possibly resized image.
+
+        Determine how the camera intrinsics and images should be jointly rescaled based on desired img. resolution.
+        Each loader implementation should set a `_max_resolution` attribute.
+
+        Args:
+            the index to fetch.
+
+        Returns:
+            intrinsics for the given camera.
+        """
+        intrinsics_full_res = self.get_camera_intrinsics_full_res(index)
+        if intrinsics_full_res is None:
+            raise ValueError(f"No intrinsics found for index {index}.")
```

Comment:
Why raise a value error? We should support datasets with no intrinsics too.

Reply:
If we don't raise an error here, we'll crash in an opaque error message later in this function.

I think every loader must return intrinsics (EXIF, or from GT, or from COLMAP, or as a rough guess based on some function of the image size). But this code would crash without it.